feat: support for SCSS files in Requirements::themedCSS()

gurucomkz opened the pull request:
I felt the need for this for a while.

Adds support for the following code:

```php
Requirements::themedCSS('css/main.scss');
Requirements::themedCSS('css/main');
```

```php
<?php
use SilverStripe\View\Requirements;
use Axllent\Scss\ScssCompiler;

// ...
Requirements::combine_files('combined.css', [
    ScssCompiler::findThemedSCSS('css/stylesheet'),
    ScssCompiler::findThemedSCSS('css/colours'),
]);
Requirements::process_combined_files();
```

```
<% require themedCSS(css/stylesheet.scss) %>
```

axllent commented:
@gurucomkz That's awesome, I love it, thank you! I'm having a few issues though (and not much time to debug right now unfortunately). My file is `themes/site/css/bootstrap.scss`, and I have:
```yaml
SilverStripe\View\SSViewer:
  themes:
    - '$public'
    - 'site'
    - '$default'
```
But:
- `ScssCompiler::findThemedSCSS('css/bootstrap');` works
- `ScssCompiler::findThemedSCSS('css/bootstrap.scss');` does not
- `ScssCompiler::findThemedSCSS('bootstrap');` does not
- `ScssCompiler::findThemedSCSS('bootstrap.scss');` does not
- `Requirements::themedCSS('css/bootstrap');` does not
- `Requirements::themedCSS('css/bootstrap.scss');` does not
- `<% require themedCSS(css/bootstrap) %>` works
- `<% require themedCSS(css/bootstrap.scss) %>` does not
- `<% require themedCSS(bootstrap) %>` does not
- `<% require themedCSS(bootstrap.scss) %>` does not
What am I missing?

gurucomkz commented:
@axllent 
Thank you for checking this out!
Some of your use cases would not work because the lookup for the files is done in `scss` folder (I just thought it would make more sense to separate CSS from SCSS files) :
* `ScssCompiler::findThemedSCSS('bootstrap');`
* `ScssCompiler::findThemedSCSS('bootstrap.scss');`
* `<% require themedCSS(bootstrap) %>`
* `<% require themedCSS(bootstrap.scss) %>`

The rest must work now - it was my silly mistake to blame.

axllent commented:
Thanks. I think the static `findThemedSCSS()` has the same `substr($name, -4)` issue which I think will break too (I haven't tested your updated code, but I have read the code). If you could please fix that then I'll have another look tomorrow evening?

I also have two another questions / comments in the meantime:

Based on your code, I see that you have a hardcoded default lookup in a `scss` folder, with a fallback to the provided path. This means that for anyone who does not use the very shorthand code (eg: `Requirements::themedCSS('bootstrap')` or `Requirements::themedCSS('bootstrap')`) and have a `<$themeDir>/scss/` folder, there is an unnecessary lookup for each file with every page load.

This is bad practice, and also not "standard" with the "standard Silverstripe way", so I would also actually ask you remove this `scss/*` lookup entirely and just use `Requirements::themedCSS('scss/bootstrap')` like you would with the default CSS classes in Silverstripe. Whilst it might be handy for you to shorthand the code in your controller, it's really a personal "feature" (to you) which I do not think the `scss` path should be hardcoded in this module.

Lastly, `ScssCompiler::findThemedSCSS()` is a public static function (also referenced in your examples). What purpose does it serve to have that second static method available? I ask because, from experience, if we provide two separate public ways to include SCSS files, users will use one or the other, or even both. Then I have to continue to provide these two methods, and maintain both methods, or risk breaking websites if `findThemedSCSS` ever changed. What is wrong with only providing the standard public `Requirements::themedCSS()` method which is identical to the existing functionality?

gurucomkz commented:
Sure, I can remove `scss` folder lookup.

#### ...for anyone who does not use the very shorthand code... there is an unnecessary lookup for each file with every page load.

Currently my lookups are done in the following order:
1. givenfile.css
2. css/givenfile.css
3. givenfile.scss
4. scss/givenfile.scss

First two are inherited from the silverstripe core.
We can speculate here that if people have installed this extension they expect using the scss files by default. In this case we can swap steps 1+2 and 3+4 and search for SCSS files first. We can go even further for those who decides include a file extension every time - we can check for extension and perform only the relevant half of the lookups.

#### RE `Requirements::themedCSS()` looking similar to `ScssCompiler::findThemedSCSS()`:

`Requirements::themedCSS()` invokes a `Requirements_Backend` via Injector which allowed you to write this module in the first place.
Unfortunately, to let my code work we also would need to replace `ThemeResourceLoader`, but that one is invoked via internal `::inst()` and cannot be replaced that easily. That is why I created a copy of `ThemeResourceLoader::findThemedCSS` as `findThemedSCSS` in the first place. 

Then I wanted to make use of the files combining:
```
Requirements::combine_files('combined.css', [
    ScssCompiler::findThemedSCSS('css/stylesheet'),
    ScssCompiler::findThemedSCSS('css/colours'),
]);
```
This works only when the files had already been resolved to their themed version, right? For that I made the resolver publicly available.

BTW, I can't find any other `substr($name, -4)` issue, sorry.

gurucomkz commented:
@axllent Sorry, I revisited the code and yes, your statement "hardcoded default lookup in a scss folder, with a fallback to the provided path." is correct. I did that because my `findThemedSCSS` is a just a minor rewrite of the original `ThemeResourceLoader::findThemedCSS` which does the same, only for the `css/` folder.

The guys from SilverStripe do it that way - I assume their logic is if you are using a certain type of a file, you must have placed it in a proper place and there is no need to explicitly say that those files are there, the lookup will pick it up.
That is exactly the reason I decided to use "scss" folder in my function - because it's a different type of files and it belongs to a different folder.

axllent commented on docs/en/Usage.md:
````diff
@@ -20,6 +20,13 @@ class PageController extends ContentController
 }
 ```
 
+The library supports `themedCSS()` file resolving mechanism. The following 3 liines are equivalent:
````
Typo in "liines" ;-)

axllent commented:
I now understand what you based your methods on, which all makes much more sense now. Also my apologies about my last comment regarding `substr($name, -4)` too - I was reading the latest [patch](https://github.com/axllent/silverstripe-scss/pull/4.patch) and didn't realise that the patch itself is split into two parts (for both your commits), and I saw the part referring to the fix as a separate instance of `substr()`, with the original `-4` part also included in the patch (so the patch actually patches itself) ;-)

I'm really surprised that `ThemeResourceLoader::findThemedCSS()` defaults lookups to `css/$name` first, but maybe that is how the majority of users use it? I had never noticed this before, probably because I always provide full paths, but if that's the way they do it then you are correct to replicate that with `scss/$name` lookup first. Quite likely we are talking fractions of milliseconds per lookup anyway, so leave it like you have :+1: 

In terms of lookup order (css first, then scss), my thinking is that anyone using this module would be using predominantly, if not solely scss files, so I would prefer the order switched (scss lookup first, followed by css). There is also a typo in your documentation (I left a comment) - not that I should make any comments about spelling as I'm renowned for typos myself ;-)

I will definitely be merging this in as I think it's a great addition, I just need to do some thorough testing first once you have made those updates please (sorry about the delays, I'm just a bit overworked at the moment haha so it may take a few more days). 

Thank you for the explanations and contributions, I really do appreciate it!

gurucomkz commented:
I appreciate your attention very much. And no worries, I'm not in a hurry - I, myself, already started using this via repo substitution. The end of the year must be one busy time for everyone :)

I fixed the typo, changed the lookup order, changed the error message for when nothing is found and added some text to the docs to, hopefully, clear this "which folder is being used by default" question.

axllent commented:
Fantastic, it's looking great, thank you @gurucomkz . I'll be in touch!

axllent commented:
@gurucomkz This has now been merged and released, together with a few other minor changes (mainly documentation) ;-) Thanks again for your contribution!
